source/common/upstream/cluster_manager_impl.cc

```diff
@@ -37,10 +37,10 @@ namespace Envoy {
 namespace Upstream {
 
 void ClusterManagerInitHelper::addCluster(Cluster& cluster) {
-  if (state_ == State::AllClustersInitialized) {
-    cluster.initialize([this, &cluster] { per_cluster_init_callback_(cluster); });
-    return;
-  }
+  // The init helper is only used during initial server load due to the overall complexity of
+  // first time initialization. After that point, the cluster manager shifts to managing warming
```

Comment:
I think I had this same question for LDS, but can you remind me why the ClusterManager can't handle both the initial server and dynamic cluster addition cases without the need for explicit initialization state here?

Reply:
It probably can with a substantial refactor, but the code is so complicated as it is that this seemed like the easier to understand way of accomplishing it. Basically, once the server is fully initialized, each cluster is warmed independently without having to worry about overall server/CM init. I will add more comments and maybe a TODO.